core/services/ocr3/plugins/ccip/commit/plugin.go

```diff
@@ -3,115 +3,594 @@ package commit
 
-	newMsgDetails := make([]model.CCIPMsgBaseDetails, 0, len(newMsgs))
-	for _, msg := range newMsgs {
-		p.lggr.Debugf("new msg: %s", msg)
-		newMsgDetails = append(newMsgDetails, msg.CCIPMsgBaseDetails)
+	newMsgs, err := p.observeNewMsgs(ctx, maxSeqNumsPerChain)
+	if err != nil {
+		return types.Observation{}, fmt.Errorf("observe new messages: %w", err)
 	}
 
-	return model.NewCommitPluginObservation(p.nodeID, newMsgDetails).Encode()
+	// TODO: The code below is related to token and gas prices and should be cleaned up in relevant PRs...
+	knownSourceChainsSlice := p.knownSourceChains.ToSlice()
+	sort.Slice(knownSourceChainsSlice, func(i, j int) bool { return knownSourceChainsSlice[i] < knownSourceChainsSlice[j] })
+
+	// Find the gas prices for each chain.
+	gasPricesVals, err := p.ccipReader.GasPrices(ctx, knownSourceChainsSlice)
+	if err != nil {
+		return nil, fmt.Errorf("get gas prices: %w", err)
+	}
+	p.lggr.Debugw("reading gas prices", "chains",
+		len(knownSourceChainsSlice), "gasPrices", len(gasPricesVals))
+	gasPrices := make([]model.GasPriceChain, 0, len(knownSourceChainsSlice))
+	for i, ch := range knownSourceChainsSlice {
+		p.lggr.Debugw("gas price", "chain", ch, "price", gasPricesVals[i])
+		gasPrices = append(gasPrices, model.NewGasPriceChain(gasPricesVals[i], ch))
+	}
+
+	// Find the token prices.
+	tokenPrices := make([]model.TokenPrice, 0) // TODO: token prices ...
+
+	p.lggr.Infow("submitting observation", "observedNewMsgs", len(newMsgs), "gasPrices", len(gasPrices), "tokenPrices", len(tokenPrices))
+	return model.NewCommitPluginObservation(newMsgs, gasPrices, tokenPrices, maxSeqNumsPerChain).Encode()
 }
 
-func (p *Plugin) ValidateObservation(outctx ocr3types.OutcomeContext, _ types.Query, ao types.AttributedObservation) error {
-	_, err := model.DecodeCommitPluginObservation(ao.Observation)
-	return err
+func (p *Plugin) ValidateObservation(_ ocr3types.OutcomeContext, _ types.Query, ao types.AttributedObservation) error {
+	obs, err := model.DecodeCommitPluginObservation(ao.Observation)
+	if err != nil {
+		return fmt.Errorf("decode commit plugin observation: %w", err)
+	}
+
+	if err := p.validateObservedSequenceNumbers(obs.NewMsgs, obs.MaxSeqNums); err != nil {
+		return fmt.Errorf("validate sequence numbers: %w", err)
+	}
+
+	if err := p.validateObserverReadingEligibility(ao.Observer, obs.NewMsgs); err != nil {
+		return fmt.Errorf("validate observer %d reading eligibility: %w", ao.Observer, err)
+	}
+
+	if err := p.validateObservedGasAndTokenPrices(obs.GasPrices, obs.TokenPrices); err != nil {
+		return fmt.Errorf("validate gas and token prices: %w", err)
+	}
+
+	return nil
 }
 
-func (p *Plugin) ObservationQuorum(outctx ocr3types.OutcomeContext, query types.Query) (ocr3types.Quorum, error) {
-	// across all chains we require at least 2f+1 observations.
+func (p *Plugin) ObservationQuorum(_ ocr3types.OutcomeContext, _ types.Query) (ocr3types.Quorum, error) {
+	// Across all chains we require at least 2f+1 observations.
```

Comment:
```suggestion
	// Across all chains we require at least 2F+1 observations.
```

Lets make it clear that this is big F not f_chain ("little f")